src/client/actions.ts

```diff
@@ -0,0 +1,133 @@
+ * wishes to receive when that event happens. This is known as the action {@link NotificationActionResult|result}.
+ *
+ * All actions passed back to an application come through a single entry point. The application should attach a
+ * listener for the {@link NotificationActionEvent|`notification-action`} event. This listener will be called once for
+ * each action raised by the service.
+ *
+ * For an overview of actions, consider the sample notification below:
+ * ```ts
+ * import {addEventListener, create} from 'openfin-notifications';
+ *
+ * // Create a notification with two buttons
+ * create({
+ *     // Basic info
+ *     title: 'Reminder',
+ *     body: 'Event "Weekly Meeting" is starting soon...',
+ *     category: 'Upcoming Events',
+ *
+ *     // We'll use the 'customData' field to store metadata about the event
+ *     customData: {eventId: '12345'},
+ *
+ *     // We want the user clicking the notification to open the associated event, so register an 'onSelect' action
+ *     onSelect: {action: 'view-event', target: 'popup'},
+ *
+ *     buttons: [
+ *         // A button that closes the notification and doesn't prompt the user about this event again. Since there is
+ *         // no special action required to handle clicking this button, there is no 'onClick' payload. This means that
+ *         // the application will not receive a "notification-action" event when the button is clicked
+ *         {
+ *             title: 'Dismiss',
+ *             iconUrl: 'https://www.example.com/cancel.png'
+ *         },
+ *
+ *         // A button that will schedule another reminder for 5 minutes from now. Since the application will need to
+ *         // perform an action on this interaction, it will need to know about the user clicking this button. By
+ *         // setting 'onClick', the service will raise a "notification-action" event when this button is clicked, and
+ *         // will pass the value of 'onClick' as the 'result' field within the event.
+ *         {
+ *             title: 'Snooze for 5 minutes',
+ *             iconUrl: 'https://www.example.com/timer.png',
+ *             onClick: {
+ *                 action: 'schedule-reminder',
+ *                 intervalMs: 5 * 60 * 1000
+ *             }
+ *         }
+ *     ]
+ * });
+ *
+ * // Create a listener that will be called for each action
+ * // Note: This handler will be used for ALL actions, from ALL notifications that are created by this application.
+ * addEventListener('notification-action', (event: NotificationActionEvent) => {
+ *     const {result, notification} = event;
+ *
+ *     if (result.action === 'view-event') {
+ *         // Open a window with full details of the associated event
+ *         openEventDetails(notification.customData.eventId, result.target);
+ *     } else if (result.action === 'schedule-reminder') {
+ *         // Schedule a new notification
+ *         scheduleReminder(notification.customData.eventId, Date.now() + result.intervalMs);
```

Comment:
Might be something we just accept, but feels weird that for the notification itself, custom data is buried below `customData`, but for the event, the custom data is just there on the top level

Reply:
Oh, wait, is `action` itself a custom field and that's just our convention?

Reply:
Not sure I fully get what you're saying...

Both `customData` and `result` are left as being app-defined, but in my mind they both exist at the same level (`NotificationOptions.customData` and `NotificationActionEvent.result`).

The latter does allow apps to specify typing though (see doc comment on `NotificationActionResult`), which is a little inconsistent with `customData` - not sure if allowing the same on custom data makes sense.

Reply:
Resolving as my confusion is better captured above!